### PR TITLE
feat(tasks): migration for 0.12.x

### PR DESCRIPTION
**Pull request by TorstenDittmann**
## What does this PR do?

Migrates the old DB to the new one:
- gives the option to migrate:
  - complete database
  - only collections
  - no database
- converts nested documents to JSON strings
- converts all numeric values and attributes to floats
- migrates all permission changes
- migrates individual changes to documents of specific collections

https://user-images.githubusercontent.com/1759475/145567025-94d88586-73c9-40c2-9e10-02c8cb393adb.mov

## Test Plan

> not worth it

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

✅

### CODE REVIEW

File: src/Appwrite/Migration/Version/V11.php [hunk 1]
```diff
@@ -0,0 +1,683 @@
...
+
+    public function __construct(PDO $db, Redis $cache = null, array $options = [])
+    {
+        parent::__construct($db, $cache, $options);
+        $this->options = array_map(fn ($option) => $option === 'yes' ? true : false, $this->options);
+
+        if (!is_null($cache)) {
+            $cacheAdapter = new Cache(new RedisCache($this->cache));
+            $this->dbInternal = new Database(new MariaDB($this->db), $cacheAdapter);
+            $this->dbExternal = new Database(new MariaDB($this->db), $cacheAdapter);
```
Comment (kodumbeats) [inline src/Appwrite/Migration/Version/V11.php:47]: I had to go looking to see if `$project` was in scope, so I think a comment helps here.
```suggestion
            $this->dbInternal = new Database(new MariaDB($this->db), $cacheAdapter); // namespace is set on execution
            $this->dbExternal = new Database(new MariaDB($this->db), $cacheAdapter); // namespace is set on execution
```

**Review by kodumbeats — APPROVED** [review]
I just had small questions, mostly. LGTM 👍🏻

File: src/Appwrite/Migration/Migration.php [hunk 2]
```diff
@@ -117,7 +142,7 @@ public function forEachDocument(callable $callback): void
                         }
 
                         try {
-                            $new = $this->projectDB->overwriteDocument($document->getArrayCopy());
+                            $new = $this->projectDB->overwriteDocument($new->getArrayCopy());
```
[the remark this answers is not in the document]
Comment (TorstenDittmann) [reply src/Appwrite/Migration/Migration.php:145]: It was passed by reference and worked - I just made it explicit :)

File: app/tasks/migrate.php [hunk 3]
```diff
@@ -11,63 +11,110 @@
...
+            Console::info("Do you want to migrate your Database Collections?");
+            $options['migrateCollections'] = Console::confirm("Type 'yes' or 'no':");
+
+            if ($options['migrateCollections'] === 'yes') {
+                Console::log('');
+                Console::log('Documents');
+                Console::log('------------------');
+                Console::warning('Be aware that following actions will happen during the migration:');
+                Console::warning('- Nested Documents will be stored as JSON values');
+                Console::warning('- All Numeric values will be converted to float');
```
[the remark this answers is not in the document]
Comment (TorstenDittmann) [reply app/tasks/migrate.php:56]: Added more informations 👍🏻

**Review by kodumbeats — APPROVED** [review]
LGTM 👍🏻

**Pull request merged**